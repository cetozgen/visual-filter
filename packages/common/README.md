# visual-filter/common

Shared utilities for the @visual-filter packages
