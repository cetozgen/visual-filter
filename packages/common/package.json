{
  "name": "@visual-filter/common",
  "version": "1.0.3",
  "description": "Shared utilities for the @visual-filter packages",
  "keywords": [
    "visual-filter"
  ],
  "homepage": "https://github.com/obadakhalili/vue-visual-filter.git",
  "repository": {
    "type": "git",
    "url": "https://github.com/obadakhalili/vue-visual-filter.git"
  },
  "license": "MIT",
  "author": "Obada Khalili",
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "clean": "rm -rf dist",
    "dev": "yarn build --watch",
    "format": "yarn run prettier --ignore-path ../../.gitignore --write",
    "preversion": "yarn build"
  },
  "devDependencies": {
    "prettier": "2.7.1",
    "typescript": "4.7.4"
  }
}
