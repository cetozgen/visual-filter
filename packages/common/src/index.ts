export enum FilterType {
  GROUP = "group",
  CONDITION = "condition",
}

export enum GroupType {
  AND = "and",
  NOT_AND = "not and",
  OR = "or",
  NOT_OR = "not or",
}

export enum DataType {
  NUMERIC = "numeric",
  NOMINAL = "nominal",
}

export function deepCopy(src: any): any {
  if (Object(src) !== src) {
    return src
  }
  if (src.constructor === Object) {
    return Object.entries(src).reduce(
      (objCopy, [key, value]) => ({ ...objCopy, [key]: deepCopy(value) }),
      {},
    )
  }
  return src.map(deepCopy)
}
