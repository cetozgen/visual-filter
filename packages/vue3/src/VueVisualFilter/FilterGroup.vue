<script>
export default {
  name: "FilterGroup",
  emits: ["addFilter", "deleteGroup"],
  props: {
    group: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      },
    },
    filterTypes: {
      type: Array,
      required: true,
    },
    groupTypes: {
      type: Array,
      required: true,
    },
    removable: {
      type: Boolean,
      required: true,
    },
  },
  methods: {
    addFilter(newFilterType) {
      if (this.filterTypes.includes(newFilterType)) {
        this.$emit("addFilter", this.group.filters, newFilterType)
      }
    },
  },
}
</script>

<template>
  <div class="space-y-4">
    <div class="space-x-2">
      <slot name="groupTypes" v-bind="{ groupTypes, group }">
        <select v-model="group.groupType">
          <option v-for="type in groupTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot name="filterAddition" v-bind="{ filterTypes, addFilter }">
        <select @change="addFilter($event.target.value)">
          <option v-for="type in filterTypes" :key="type" :value="type">
            {{ type }}
          </option>
        </select>
      </slot>
      <slot
        v-if="removable"
        name="groupDeletion"
        :deleteGroup="() => $emit('deleteGroup', group)"
      >
        <button @click="$emit('deleteGroup', group)">x</button>
      </slot>
    </div>
    <div v-if="group.filters.length" class="ml-10 space-y-1">
      <slot name="groupChildren"></slot>
    </div>
  </div>
</template>
