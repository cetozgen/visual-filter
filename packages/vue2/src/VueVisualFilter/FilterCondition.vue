<script>
import { DataType } from "@visual-filter/common"

export default {
  name: "FilterCondition",
  emits: ["updateField", "deleteCondition"],
  props: {
    condition: {
      type: Object,
      required: true,
      validator(value) {
        return value.constructor === Object
      },
    },
    fieldNames: {
      type: Array,
      required: true,
    },
    numericMethodNames: {
      type: Array,
      required: true,
    },
    nominalMethodNames: {
      type: Array,
      required: true,
    },
  },
  computed: {
    isNumeric() {
      return this.condition.dataType === DataType.NUMERIC
    },
  },
  methods: {
    updateField(newFieldName) {
      if (this.fieldNames.includes(newFieldName)) {
        this.$emit("updateField", this.condition, newFieldName)
      }
    },
  },
}
</script>

<template>
  <div class="space-x-2">
    <slot name="fieldUpdation" v-bind="{ fieldNames, condition, updateField }">
      <select
        v-model="condition.fieldName"
        @change="updateField($event.target.value)"
      >
        <option v-for="field in fieldNames" :key="field" :value="field">
          {{ field }}
        </option>
      </select>
    </slot>
    <slot
      name="methodUpdation"
      v-bind="{
        numericMethodNames: isNumeric && numericMethodNames,
        nominalMethodNames: isNumeric || nominalMethodNames,
        condition,
      }"
    >
      <select v-model="condition.method">
        <option
          v-for="method in isNumeric ? numericMethodNames : nominalMethodNames"
          :key="method"
          :value="method"
        >
          {{ method }}
        </option>
      </select>
    </slot>
    <slot name="argumentUpdation" :condition="condition">
      <input type="text" v-model="condition.argument" />
    </slot>
    <slot
      name="conditionDeletion"
      :deleteCondition="() => $emit('deleteCondition', condition)"
    >
      <button @click="$emit('deleteCondition', condition)">x</button>
    </slot>
  </div>
</template>
