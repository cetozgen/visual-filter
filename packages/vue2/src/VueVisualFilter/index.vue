<script>
import {
  FilterType,
  GroupType,
  DataType,
  deepCopy,
} from "@visual-filter/common"
import applyFilter from "@visual-filter/applyer"

import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"

export default {
  name: "VueVisualFilter",
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        console.log("in validation",value);
        try {
          return (
            value.data.length &&
            value.data.every(
              (field, index, fields) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array &&
                (index > 0
                  ? true
                  : true),
            ) &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function",
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function",
            )
          )
        } catch {
          return false
        }
      },
    },
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: [],
      },
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    },
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        if (this.$listeners["filter-update"]) {
          this.$emit("filter-update", {
            filter: deepCopy(this.filter),
            data: applyFilter(
              this.filter,
              this.filteringOptions.methods,
              deepCopy(this.filteringOptions.data),
            ),
          })
        }
      },
    },
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      console.log("in updateConditionField", condition, newFieldName);
      let newSampleValues = [];
      const {
        type: newType,
        values: [newSampleValue = ""],
      } = this.filteringOptions.data.find(
        (field) => {
          field.name === newFieldName
        },
      )
      if (condition.dataType !== newType) {
        condition.method =
          (newType === DataType.NUMERIC
            ? this.numericMethodNames[0]
            : this.nominalMethodNames[0]) || ""
        condition.argument = newSampleValues
        condition.dataType = newType
      }
    },
    addFilter(filters, newFilterType) {
      console.log("in addFilter",newFilterType);
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: [],
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue = ""],
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            (type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0]) || "",
          argument: sampleValue,
        })
      }
    },
    deleteFilter(filterIndex, parentGroup) {
      parentGroup.splice(filterIndex, 1)
    },
  },
  render(createElement) {
    return createVisualFilter.call(this, this.filter)

    function createVisualFilter(filter, filterIndex, parentGroup) {
      if (filter.type === FilterType.GROUP) {
        const removable = filter !== this.filter

        return createElement(FilterGroup, {
          props: {
            group: filter,
            removable,
          },
          on: {
            addFilter: this.addFilter,
            deleteGroup: removable
              ? () => this.deleteFilter(filterIndex, parentGroup)
              : () => {},
          },
          scopedSlots: {
            groupTypes: this.$scopedSlots.groupTypes,
            filterAddition: this.$scopedSlots.filterAddition,
            groupDeletion: this.$scopedSlots.groupDeletion,
            groupChildren: () =>
              filter.filters.map((childFilter, childFilterIndex, parentGroup) =>
                createVisualFilter.call(
                  this,
                  childFilter,
                  childFilterIndex,
                  parentGroup,
                ),
              ),
          },
        })
      } else {
        return createElement(FilterCondition, {
          props: {
            condition: filter,
            fieldNames: this.fieldNames,
            numericMethodNames: this.numericMethodNames,
            nominalMethodNames: this.nominalMethodNames,
          },
          on: {
            updateField: this.updateConditionField,
            deleteCondition: () => this.deleteFilter(filterIndex, parentGroup),
          },
          scopedSlots: {
            fieldUpdation: this.$scopedSlots.fieldUpdation,
            methodUpdation: this.$scopedSlots.methodUpdation,
            argumentUpdation: this.$scopedSlots.argumentUpdation,
            conditionDeletion: this.$scopedSlots.conditionDeletion,
          },
        })
      }
    }
  },
}
</script>
