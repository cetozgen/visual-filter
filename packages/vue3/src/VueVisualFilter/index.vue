<script>
import { h } from "vue"
import {
  FilterType,
  GroupType,
  DataType,
  deepCopy,
} from "@visual-filter/common"
import applyFilter from "@visual-filter/applyer"

import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"

export default {
  name: "VueVisualFilter",
  emits: ["filterUpdate"],
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        try {
          return (
            value.data.length &&
            value.data.every(
              (field, index, fields) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array &&
                (index > 0
                  ? field.values.length === fields[index - 1].values.length
                  : true),
            ) &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function",
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function",
            )
          )
        } catch {
          return false
        }
      },
    },
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: [],
      },
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    },
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        this.$emit("filterUpdate", {
          filter: deepCopy(this.filter),
          data: applyFilter(
            this.filter,
            this.filteringOptions.methods,
            deepCopy(this.filteringOptions.data),
          ),
        })
      },
    },
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      const {
        type: newType,
        values: [newSampleValue = ""],
      } = this.filteringOptions.data.find(
        (field) => field.name === newFieldName,
      )
      if (condition.dataType !== newType) {
        condition.method =
          (newType === DataType.NUMERIC
            ? this.numericMethodNames[0]
            : this.nominalMethodNames[0]) || ""
        condition.argument = newSampleValue
        condition.dataType = newType
      }
    },
    addFilter(filters, newFilterType) {
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: [],
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue = ""],
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            (type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0]) || "",
          argument: sampleValue,
        })
      }
    },
    deleteFilter(filterToDelete) {
      function recursiveDeletion(filter, index, filters) {
        if (filter === filterToDelete) {
          filters.splice(index, 1)
        } else if (filter.type === FilterType.GROUP) {
          filter.filters.map(recursiveDeletion)
        }
      }

      if (filterToDelete !== this.filter) {
        recursiveDeletion(this.filter)
      }
    },
  },
  render() {
    const createVisualizer = (filter) => {
      if (filter.type === FilterType.GROUP) {
        return h(
          FilterGroup,
          {
            group: filter,
            filterTypes: Object.values(FilterType),
            groupTypes: Object.values(GroupType),
            removable: filter !== this.filter,
            onAddFilter: this.addFilter,
            onDeleteGroup: this.deleteFilter,
          },
          {
            groupTypes: this.$slots.groupTypes,
            filterAddition: this.$slots.filterAddition,
            groupDeletion: this.$slots.groupDeletion,
            groupChildren: () => filter.filters.map(createVisualizer),
          },
        )
      } else {
        return h(
          FilterCondition,
          {
            condition: filter,
            fieldNames: this.fieldNames,
            numericMethodNames: this.numericMethodNames,
            nominalMethodNames: this.nominalMethodNames,
            onUpdateField: this.updateConditionField,
            onDeleteCondition: this.deleteFilter,
          },
          {
            fieldUpdation: this.$slots.fieldUpdation,
            methodUpdation: this.$slots.methodUpdation,
            argumentUpdation: this.$slots.argumentUpdation,
            conditionDeletion: this.$slots.conditionDeletion,
          },
        )
      }
    }

    return createVisualizer(this.filter)
  },
}
</script>
