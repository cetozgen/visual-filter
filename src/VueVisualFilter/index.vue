<script>
import { h } from "vue"
import FilterGroup from "./FilterGroup.vue"
import FilterCondition from "./FilterCondition.vue"
import { createEnum, deepCopy, applyFilter } from "@/helpers.js"

export const FilterType = createEnum({
  GROUP: "group",
  CONDITION: "condition"
})

export const GroupType = createEnum({
  AND: "and",
  NOT_AND: "not and",
  OR: "or",
  NOT_OR: "not or"
})

export const DataType = createEnum({
  NUMERIC: "numeric",
  NOMINAL: "nominal"
})

export default {
  name: "VisualFilter",
  emits: ["filterUpdate"],
  props: {
    filteringOptions: {
      type: Object,
      required: true,
      validator(value) {
        try {
          return (
            value.data.length &&
            value.data.every(
              (field) =>
                typeof field.name === "string" &&
                typeof field.type === "string" &&
                field.values.constructor === Array
            ) &&
            value.methods.numeric &&
            Object.values(value.methods.numeric).every(
              (method) => typeof method === "function"
            ) &&
            Object.values(value.methods.nominal).every(
              (method) => typeof method === "function"
            )
          )
        } catch {
          return false
        }
      }
    }
  },
  data() {
    return {
      filter: {
        type: FilterType.GROUP,
        groupType: GroupType.AND,
        filters: []
      }
    }
  },
  computed: {
    fieldNames() {
      return this.filteringOptions.data.map((field) => field.name)
    },
    numericMethodNames() {
      return Object.keys(this.filteringOptions.methods.numeric)
    },
    nominalMethodNames() {
      return Object.keys(this.filteringOptions.methods.nominal)
    }
  },
  watch: {
    filter: {
      deep: true,
      handler() {
        this.$emit("filterUpdate", {
          filter: deepCopy(this.filter),
          data: applyFilter(
            this.filter,
            this.filteringOptions.methods,
            deepCopy(this.filteringOptions.data)
          )
        })
      }
    }
  },
  methods: {
    updateConditionField(condition, newFieldName) {
      const {
        type,
        values: [sampleValue]
      } = this.filteringOptions.data.find(
        (field) => field.name === newFieldName
      )
      condition.fieldName = newFieldName
      condition.dataType = type
      condition.method =
        type === DataType.NUMERIC
          ? this.numericMethodNames[0]
          : this.nominalMethodNames[0]
      condition.argument = sampleValue
    },
    updateConditionMethod(condition, newMethodName) {
      condition.method = newMethodName
    },
    updateConditionArgument(condition, newArgumentValue) {
      condition.argument = newArgumentValue
    },
    updateGroupType(group, newGroupType) {
      group.groupType = newGroupType
    },
    addFilter(filters, newFilterType) {
      if (newFilterType === FilterType.GROUP) {
        filters.push({
          type: FilterType.GROUP,
          groupType: GroupType.AND,
          filters: []
        })
      } else {
        const {
          name,
          type,
          values: [sampleValue]
        } = this.filteringOptions.data[0]

        filters.push({
          type: FilterType.CONDITION,
          fieldName: name,
          dataType: type,
          method:
            type === DataType.NUMERIC
              ? this.numericMethodNames[0]
              : this.nominalMethodNames[0],
          argument: sampleValue
        })
      }
    },
    deleteFilter(filterToDelete) {
      function recursiveDeletion(filter, index, filters) {
        if (filter === filterToDelete) {
          filters.splice(index, 1)
        } else if (filter.type === FilterType.GROUP) {
          filter.filters.map(recursiveDeletion)
        }
      }

      if (filterToDelete !== this.filter) {
        recursiveDeletion(this.filter)
      }
    }
  },
  render() {
    const createVisualizer = (filter) => {
      if (filter.type === FilterType.GROUP) {
        return h(
          FilterGroup,
          {
            group: filter,
            filterTypes: Object.values(FilterType),
            groupTypes: Object.values(GroupType),
            removable: filter !== this.filter,
            onUpdateType: this.updateGroupType,
            onAddFilter: this.addFilter,
            onDeleteGroup: this.deleteFilter
          },
          () => filter.filters.map(createVisualizer)
        )
      } else {
        return h(FilterCondition, {
          condition: filter,
          fieldNames: this.fieldNames,
          numericMethodNames: this.numericMethodNames,
          nominalMethodNames: this.nominalMethodNames,
          onUpdateField: this.updateConditionField,
          onUpdateMethod: this.updateConditionMethod,
          onUpdateArgument: this.updateConditionArgument,
          onDeleteCondition: this.deleteFilter
        })
      }
    }

    return createVisualizer(this.filter)
  }
}
</script>
