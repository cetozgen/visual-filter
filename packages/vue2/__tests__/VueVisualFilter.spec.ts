import { mount } from "@vue/test-utils"
import { FilterType } from "@visual-filter/common"

import VueVisualFilter from "../src/VueVisualFilter/index.vue"

const propsData = {
  filteringOptions: {
    data: [
      { name: "First Name", type: "nominal", values: [] },
      { name: "Last Name", type: "nominal", values: [] },
      { name: "Grade", type: "numeric", values: [] },
    ],
    methods: {
      numeric: { "=": () => {}, ">": () => {}, "<": () => {} },
      nominal: { contains: () => {}, startsWith: () => {}, endsWith: () => {} },
    },
  },
}
const wrapper = mount<VueVisualFilter & { [key: string]: any }>(
  VueVisualFilter,
  { propsData },
)

describe("computed properties", () => {
  it("computes correct values", () => {
    expect(wrapper.vm.fieldNames).toMatchSnapshot()
    expect(wrapper.vm.numericMethodNames).toMatchSnapshot()
    expect(wrapper.vm.nominalMethodNames).toMatchSnapshot()
  })
})

describe("methods", () => {
  it("adds correct filter objects by addFilter method", () => {
    wrapper.vm.addFilter(wrapper.vm.filter.filters, FilterType.GROUP)
    wrapper.vm.addFilter(wrapper.vm.filter.filters, FilterType.CONDITION)
    expect(wrapper.vm.filter.filters[0]).toMatchSnapshot()
    expect(wrapper.vm.filter.filters[1]).toMatchSnapshot()
  })

  it("removes correct filter object by removeFilter method", () => {
    wrapper.vm.deleteFilter(0, wrapper.vm.filter.filters)
    expect(wrapper.vm.filter.filters).toMatchSnapshot()
  })

  it("doesn't update condition field when changed to same type", () => {
    wrapper.vm.updateConditionField(
      wrapper.vm.filter.filters[0],
      propsData.filteringOptions.data[1].name,
    )
    expect(wrapper.vm.filter.filters[0]).toMatchSnapshot()
  })

  it("updates condition field when changed to different type", () => {
    wrapper.vm.updateConditionField(
      wrapper.vm.filter.filters[0],
      propsData.filteringOptions.data[2].name,
    )
    expect(wrapper.vm.filter.filters[0]).toMatchSnapshot()
  })
})
