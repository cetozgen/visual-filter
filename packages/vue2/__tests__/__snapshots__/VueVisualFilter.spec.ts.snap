// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`computed properties computes correct values 1`] = `
Array [
  "First Name",
  "Last Name",
  "Grade",
]
`;

exports[`computed properties computes correct values 2`] = `
Array [
  "=",
  ">",
  "<",
]
`;

exports[`computed properties computes correct values 3`] = `
Array [
  "contains",
  "startsWith",
  "endsWith",
]
`;

exports[`methods adds correct filter objects by addFilter method 1`] = `
Object {
  "filters": Array [],
  "groupType": "and",
  "type": "group",
}
`;

exports[`methods adds correct filter objects by addFilter method 2`] = `
Object {
  "argument": "",
  "dataType": "nominal",
  "fieldName": "First Name",
  "method": "contains",
  "type": "condition",
}
`;

exports[`methods doesn't update condition field when changed to same type 1`] = `
Object {
  "argument": "",
  "dataType": "nominal",
  "fieldName": "First Name",
  "method": "contains",
  "type": "condition",
}
`;

exports[`methods removes correct filter object by removeFilter method 1`] = `
Array [
  Object {
    "argument": "",
    "dataType": "nominal",
    "fieldName": "First Name",
    "method": "contains",
    "type": "condition",
  },
]
`;

exports[`methods updates condition field when changed to different type 1`] = `
Object {
  "argument": "",
  "dataType": "numeric",
  "fieldName": "First Name",
  "method": "=",
  "type": "condition",
}
`;
